### ISSUE TITLE
[Strix 18/20] use geometric KV reserve for repeated Vulkan sequence-prefill chunks

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12548]
# Strix P1: use geometric KV reserve for repeated Vulkan sequence-prefill chunks

Parent campaign: #11572. Frozen comparison ladder: #12093. Matched physical comparison gate: #12098.

## Goal

Replace the exact-capacity Vulkan sequence-KV pre-reserve with one guarded geometric reserve policy for Kraw, K, and V.

## Context

- **For:** fak-native chunked Qwen3.8 prompt prefill on Strix Halo.
- **Problem:** every sequence-prefill call reserves exactly `StartPos + tokens`, so a stream of prompt chunks can reallocate and recopy all prior K, Kraw, and V data at each growth boundary.
- **Today:** `internal/compute/vulkan_qwen35_sequence.go:418-435` allocates exact `need` for all three attention planes and copies `cache.len`; the existing geometric `growAppend` policy at `internal/compute/vulkan_kv_vulkan.go:31-48` is bypassed by that pre-reserve.
- **Better because:** a bounded geometric capacity policy amortizes allocation and device-to-device copies across repeated chunks without changing logical KV length or attention semantics.
- **Witness:** deterministic capacity/copy-count assertions plus a physical repeated-chunk ablation show identical logits and >=5% median sequence-prefill improvement.

## Known comparison context (not a victory claim)

Halogen reports HIP prefill at 620/710/566 tok/s (pp512/pp2048/pp32768) and DFlash2 decode at 31.71 tok/s mean. nabe reports llama.cpp Vulkan+MTP aggregate decode from 16.65 tok/s at c1 to 30.61 tok/s at c16. The latest sealed local fak receipt is 0.34 tok/s prefill and 0.37 tok/s decode and is **NOT_COMPARABLE**. Allocation-copy savings do not establish a matched throughput win; #12098 owns that promotion.

## Source-pinned prior art and self-query

- Comparison study: `study_fea6c1bc31f94fcee01ec566289e7ca509c28e7a982068dbde17b36e33b333e3`.
- Vulkan frontier study: `study_a8e74aeef35f4f79a52d2af0a1d1f6ef8c4830eabdf419db4ddf3a2052c08c1a`.
- Pinned internal prior art: `internal/compute/vulkan_kv_vulkan.go:31-48` already implements geometric append growth for the decode KV store; this issue adapts that owned mechanism rather than importing an external allocator.
- Self-query status: **PARTIAL**. Sequence prefill has an exact-capacity reservation for correctness, while #12397 tunes chunk selection from physical duration feedback and does not own allocation-capacity growth.
- Route: **BORROW** fak's existing `growAppend` amortization rule and **ADAPT** it to three-plane sequence reservation with overflow and maximum-capacity guards.

## Dependencies and pickup

- Coordinates with: #11572, #12093, #12397
- Promotion requires: #12098

This issue has an executable allocation-policy slice now; coordination does not block pickup.

## One bounded deliverable

Replace exact-capacity sequence KV growth with one guarded geometric reserve helper shared by Kraw/K/V, preserving logical lengths and copying each old allocation at most logarithmically over repeated chunks.

Write surface (maximum two closely related files):

- `internal/compute/vulkan_qwen35_sequence.go`
- `internal/compute/vulkan_qwen35_sequence_test.go`

## Deterministic software witness

`go test -tags vulkan ./internal/compute -run TestQwen35VulkanSequenceKVReserveGeometric -count=1`

The test must feed multiple chunk sizes, assert logical positions, capacity bounds, and D2D copy counts, and compare final attention output with exact-growth reference output.

## Strix physical witness and promotion rule

`fak validate --strix --subkernels=qwen35-sequence-kv --ablate=geometric-reserve`

Capture at least five interleaved baseline/candidate samples for the same repeated-chunk schedule and compare medians. Promotion requires identical artifact, prompt chunks, driver, clocks/governor and workload; quality parity; `fallback_count=0`; engine `fak-native`; and >=5% median sequence-prefill improvement. Also report allocation and D2D byte counts.

## Native execution fence

The KV store, sequence path, and measurement remain fak-native Vulkan; no comparator runtime may service the request.

**Comment by anthony-chaudhary** [comment]
Implementation update (source-bound, no performance claim):

- The two-file candidate now uses guarded geometric reserve for Kraw/K/V, saturating at the device limit to avoid repeated near-cap copies. The deterministic test covers chunked positions, capacity bounds, modeled D2D copy counts, exact-growth attention parity, overflow, and near-cap behavior.
- Software witness is GREEN: `go test -tags vulkan ./internal/compute -run TestQwen35VulkanSequenceKVReserveGeometric -count=1` (`ok`, actual WSL-built Vulkan shim); tagged `go vet ./internal/compute` is also green.
- The required physical command is not runnable against the current validator contract: `qwen35-sequence-kv` is not a registered Strix subkernel selector, and `geometric-reserve` is not a registered ablation. The exact source-bound invocation returned `unknown subkernel selector(s): qwen35-sequence-kv` before device execution.
- Therefore there is no N>=5 interleaved physical receipt, allocation/D2D-byte report, >=5% median proof, or supported `fallback_count=0` promotion claim. The candidate remains uncommitted and unpushed per the mandatory hardware gate.

Next checkable step: extend/amend the issue write boundary to include the dedicated Strix selector + ablation/metrics receipt path, or supply an already-registered equivalent witness command.

**Comment by anthony-chaudhary** [comment]
Resuming source-bound verification at f1bcc34e0. Geometric reserve is already landed (eda91ad070, with subsequent COW integration), so this task is strengthening the existing test rather than duplicating implementation. Scope is the existing geometric-reserve test: three-plane preservation plus observed Vulkan reserve-copy counts/bytes. Coordinated disjoint test hunks with #12549. Physical qualification remains separate and pending #12611 and the matched ablation gate.

**Comment by anthony-chaudhary** [comment]
Candidate verification update: the existing reserve implementation was retained. The test now covers Kraw/K/V, brackets reserve calls with real Vulkan execution counters, records pointer-observed buffer allocations and D2D bytes, preserves logical lengths, and verifies exact data/finite attention parity. Independent review confirmed the counter and cleanup boundaries. Tagged WSL compilation and host-policy subtests pass, but DeviceKVReserveAndOutputParity SKIPs because the WSL runner has no Vulkan backend; therefore no observed device counts or performance gain are claimed. The active hardware coordinator confirmed that no authorized standalone test window is available. Physical acceptance stays open, including real sequence continuation, N>=5 interleaved timing and the >=5% threshold. No commit or closure yet; isolated validation is being retried with its supported longer timeout after the first build phase timed out.

**Comment by anthony-chaudhary** [comment]
Final software checkpoint: tagged compute test exits 0 with host policy/guard PASS and explicit device subtest SKIP; tagged compute vet exits 0. The accepted run hashes all three candidate files unchanged before/after. Full isolated validation is nonpassing: first run TIMEOUT during build at 90s, supported 4m retry failed WSL_CAPABILITY_PREFLIGHT_FAILED before workspace allocation (14.5s). No gate was skipped. Candidate remains uncommitted in its isolated task checkout, with separate recovery patch; this issue and #12611 remain OPEN. Next checkable step is restore the supported WSL validation route and obtain exclusive standalone hardware authority, then run source-bound parity and the repeated-sequence physical ablation.

**Comment by anthony-chaudhary** [comment]
Updated gate result: after WSL responsiveness recovered, the supported 4m isolated validation completed build, vet and selected tests successfully at base f1bcc34e0. Overall exit remains 1: automatic Strix validation refused UNSUPPORTED_VALIDATOR_BINARY; its recovery requires a current stamped WSL build with /proc/self/exe authority. No physical execution occurred. Meanwhile #12611 landed independently as a251f8d70, so our duplicate selector must not land. The stronger #12548 test-only patch is preserved under supported WIP checkpoint refs/fak/wip/codex-12548-d58a (b44cc751ff64e796a86b97bda431321673736100). Next is narrow reconciliation onto updated trunk and stamped-validator execution under exclusive hardware authority.

**Comment by anthony-chaudhary** [comment]
Integrated recovery checkpoint: test-only delta is now based on 1b833d297 and preserves the independently landed #12611 catalog/emitter. Only internal/compute/vulkan_qwen35_sequence_test.go changes; it also checks all plane readbacks against original inputs. Independent parent review found no blocking logic defect; gofmt/diff-check are clean. Supported WIP ref refs/fak/wip/codex-12548-integrated points b3f2c07998f9d7189f5fd0562cd4335910fb1059. Latest required validation is blocked by the existing #6827 WSL capability-preflight timeout; do not duplicate that issue. A fresh-shim tagged test is still running, so current-source tagged execution is pending. No commit/closure or physical promotion.

**Comment by anthony-chaudhary** [comment]
The integrated tagged witness has now finished naturally with exit 0: GeometricPolicyBoundsAndCopyReduction and OverflowAndCeilingGuards PASS; DeviceKVReserveAndOutputParity explicitly SKIPs (Vulkan backend unavailable). Test execution itself reports 0.024s after the long compilation/wait. No process kill, cache reset, or physical run occurred. Current one-file candidate remains checkpointed at b3f2c079 on base 1b833d297. Required full validation and physical acceptance remain open; no landing or performance claim.

**Comment by anthony-chaudhary** [comment]
Resolved: already implemented and green on origin/main; closed by witnessed software evidence.

Witness (SW-VERIFIED, re-run 2026-09-13T05:38Z on HEAD 85f433a2):`go test -tags vulkan ./internal/compute -run TestQwen35VulkanSequenceKVReserveGeometric -count=1 -v` -> PASS (0.21s). GeometricPolicyBoundsAndCopyReduction: 16 repeated chunks exact copies=15 vs geometric=4; mixed 8 chunks exact=7 vs geometric=5. DeviceKVReserveAndOutputParity: max_abs_delta=0, finite_output=true, logical_len=4096, exact_cap=4096, geom_cap=4096; emitted fak.strix.subkernel-parity/v1 receipt selector=qwen35-sequence-kv, engine=fak-native/vulkan, device_observed=true, passed=true.

Implementing commits (both ancestors of origin/main): eda91ad07 perf(compute): use geometric KV reserve for repeated Vulkan sequence-prefill chunks (#12548); a251f8d70 feat(amdgpu): register source-bound Strix selector for Qwen sequence KV reserve.

The bounded software deliverable (guarded geometric reserve shared by Kraw/K/V, logical lengths preserved, logarithmic old-allocation copies) is satisfied. The [HW-WITNESSED] promotion criterion remains with the #12098 matched-physical umbrella and is not claimed here.
